### PR TITLE
Support Process.Start as a different user on Unix.

### PR DESCRIPTION
**Pull request by eerhardt**
The most common use case is to de-escalate privileges from a privileged process to a normal process.

Part of #26268 

@safern and @tarekgh - I'm not sure who owns the RemoteExecution stuff, but could you guys give it a look?

/cc @arunjvs

### CODE REVIEW

File: src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs [hunk 1]
```diff
@@ -149,6 +149,7 @@ public sealed partial class RemoteInvokeOptions
         public string ExceptionFile { [System.Runtime.CompilerServices.CompilerGeneratedAttribute]get { throw null; } }
         public int ExpectedExitCode { [System.Runtime.CompilerServices.CompilerGeneratedAttribute]get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute]set { } }
         public bool Start { [System.Runtime.CompilerServices.CompilerGeneratedAttribute]get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute]set { } }
+        public bool RunAsSudo { [System.Runtime.CompilerServices.CompilerGeneratedAttribute]get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute]set { } }
```
Comment (ViktorHofer) [inline src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs]: RunAsSudo is a very Unix specific term. As you are adding it to the refs for all platforms consider rephrasing the method.

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs [hunk 2]
```diff
@@ -58,8 +58,16 @@ private static RemoteInvokeHandle RemoteInvoke(MethodInfo method, string[] args,
             if (!File.Exists(HostRunner))
                 throw new IOException($"{HostRunner} test app isn't present in the test runtime directory.");
 
-            psi.FileName = HostRunner;
-            psi.Arguments = testConsoleAppArgs;
+            if (options.RunAsSudo)
```
Comment (ViktorHofer) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: what will happen on Windows if I set `RunAsSudo` to true? I guess it will try to find an executable called sudo and will fail?

**Comment by tarekgh** [comment]
CC @joperezr and @wtgodbe as the owners of Process

File: src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs (hunk 1 above)
Comment (eerhardt) [reply src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs]: I choose that name as it matched the existing method in this library:  https://github.com/dotnet/corefx/blob/master/src/CoreFx.Private.TestUtilities/ref/CoreFx.Private.TestUtilities.cs#L16.  That method is also on all platforms.

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs (hunk 2 above)
Comment (tarekgh) [inline src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: >if (options.RunAsSudo) [](start = 12, length = 22)

I suggest testing if it is not Windows too. or make the setter of RunAsSudo to throw if set for Windows

File: src/System.Diagnostics.Process/System.Diagnostics.Process.sln [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 14
-VisualStudioVersion = 14.0.25420.1
+# Visual Studio 15
```
Comment (maryamariyan) [inline src/System.Diagnostics.Process/System.Diagnostics.Process.sln]: Is this change necessary?

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 4]
```diff
@@ -597,6 +606,91 @@ private ProcessWaitState GetWaitState()
...
+            // Fallback to heap allocations if necessary, growing the buffer until
+            // we succeed.  TryGetPasswd will throw if there's an unexpected error.
+            int lastBufLen = BufLen;
+            while (true)
+            {
+                lastBufLen *= 2;
+                byte[] heapBuf = new byte[lastBufLen];
+                fixed (byte* buf = &heapBuf[0])
+                {
+                    if (TryGetPasswd(userName, buf, heapBuf.Length, out passwd))
```
Comment (tarekgh) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: >TryGetPasswd [](start = 24, length = 12)

Is it possible TryGetPasssword fail for other reasons than not enough buffer? if so, this can cause out of memory

Comment (tarekgh) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: never mind. I see the implementation now so it return false only in case of not enough buffer

---
In reply to: [162512699](https://github.com/dotnet/corefx/pull/26431#discussion_r162512699) [](ancestors = 162512699)

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 5]
```diff
@@ -597,6 +606,91 @@ private ProcessWaitState GetWaitState()
...
+            // If the call failed because the buffer was too small, return false to 
+            // indicate the caller should try again with a larger buffer.
+            if (errorInfo.Error == Interop.Error.ERANGE)
+            {
+                passwd = null;
+                return false;
+            }
+
+            // Otherwise, fail.
+            throw new IOException(errorInfo.GetErrorMessage(), errorInfo.RawErrno);
```
Comment (tarekgh) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: >IOException [](start = 22, length = 11)

should we throw Win32Exception instead? I think we are throwing it on Linux too in some cases

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 6]
```diff
@@ -597,6 +606,91 @@ private ProcessWaitState GetWaitState()
...
 
+        private static (uint userId, uint groupId) GetUserAndGroupIds(ProcessStartInfo startInfo)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(startInfo.UserName));
+
+            (uint? userId, uint? groupId) = GetUserAndGroupIds(startInfo.UserName);
+
+            if (userId == null || groupId == null)
+            {
+                throw new Win32Exception(SR.Format(SR.UserDoesNotExist, startInfo.UserName));
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: Are there situations where userId might be non-null but groupId null?  The error message here says that the user doesn't exist, so I'm wondering if userId will be null iff groupId is null, in which case the check could be simplified to just check userId, maybe with an assert on groupId.

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 7]
```diff
@@ -304,9 +304,93 @@ public void TestStartOnOSXWithBadFormat()
...
+                    // Assert that the effective user of the child process was never 'root'
+                    // and was the real user of this process.
+                    Assert.NotEqual(0, p.ExitCode);
+                }
+
+                return 0;
+            }
+        }
+
+        public static int TestStartWithUserName_ChildProcess()
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: I'm confused by this function and how the name relates to its implementation.

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs (hunk 2 above)
Comment (danmosemsft) [reply src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: PlatformNotSupportedException...

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs [hunk 8]
```diff
@@ -597,6 +606,91 @@ private ProcessWaitState GetWaitState()
...
+            }
+
+            return (userId.Value, groupId.Value);
+        }
+
+        private unsafe static (uint? userId, uint? groupId) GetUserAndGroupIds(string userName)
+        {
+            // First try with a buffer that should suffice for 99% of cases.
+            Interop.Sys.Passwd? passwd;
+            const int BufLen = 1024;
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: consider  defining as 0 in DEBUG to force exercising the fallback path. This is a common trick.

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 7 above)
Comment (danmosemsft) [reply src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: It enables the assert on line 376.

File: src/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.cs [hunk 9]
```diff
@@ -125,6 +126,12 @@ public string WorkingDirectory
         public bool ErrorDialog { get; set; }
         public IntPtr ErrorDialogParentHandle { get; set; }
 
+        public string UserName
+        {
+            get => _userName ?? string.Empty;
+            set => _userName = value;
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.cs]: maybe use null for "no name" and throw for setting empty? I am not sure there is a concept of an empty user name

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 10]
```diff
@@ -304,9 +304,93 @@ public void TestStartOnOSXWithBadFormat()
...
+            TestStartWithUserName();
+        }
+
+        /// <summary>
+        /// Tests when running as root and starting a new process as a normal user,
+        /// the new process doesn't have elevated privileges.
+        /// </summary>
+        [Fact]
+        [OuterLoop("Needs sudo access")]
+        [Trait(XunitConstants.Category, XunitConstants.RequiresElevation)]
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: Seems ideally tests.targets would pass `-notrait requireselevation` iff it's not running with sudo access. OuterLoop and sudo seem orthogonal

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 11]
```diff
@@ -304,9 +304,93 @@ public void TestStartOnOSXWithBadFormat()
...
+                    p.WaitForExit();
+
+                    // since the process was started with the current real user, even if this test
+                    // was run with 'sudo', the child process will be run as the normal real user.
+                    // Assert that the effective user of the child process was never 'root'
+                    // and was the real user of this process.
+                    Assert.NotEqual(0, p.ExitCode);
+                }
+
+                return 0;
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: were you intending to always return 0? you are asserting the return code is 0.

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs (hunk 6 above)
Comment (arunjvs) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: +1. Also check in line 296?

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs [hunk 12]
```diff
@@ -304,9 +304,93 @@ public void TestStartOnOSXWithBadFormat()
...
+        [Fact]
+        public void TestStartWithNormalUser()
+        {
+            TestStartWithUserName();
+        }
+
+        /// <summary>
+        /// Tests when running as root and starting a new process as a normal user,
+        /// the new process doesn't have elevated privileges.
+        /// </summary>
```
Comment (arunjvs) [inline src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: Should we test that the child process when started as normal user, cannot escalate to root again? This is possible if the implementation (accidentally) used setegid() and seteuid() instead of setgid() and setuid().

File: src/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.cs (hunk 9 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.cs]: This property was copied from the [Windows specific code](https://github.com/dotnet/corefx/blob/3b2666a7aed005c98fcb4f8ca5245c9459be311f/src/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.Windows.cs#L16-L20).  I don't think it is a good idea to change the behavior.

File: src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs (hunk 2 above)
Comment (eerhardt) [reply src/CoreFx.Private.TestUtilities/src/System/Diagnostics/RemoteExecutorTestBase.Process.cs]: I'll change it to throw an exception if you try to set `RunAsSudo` on Windows.

File: src/System.Diagnostics.Process/System.Diagnostics.Process.sln (hunk 3 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/System.Diagnostics.Process.sln]: It happened automatically by VS.  I tend to not fight VS, as it is usually a losing battle.

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 7 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: This is the function that runs in the child process.  I guess I was naming it based on where its used.

I can rename it to something that more matches what it does - like `GetCurrentEffectiveUserId()`.  Would that help?

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 11 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: Yes, there are 2 processes at play here.  There's the outer process that gets invoked with `sudo`.  That's what this exit code is returning.  Then there's the inner/child process that gets invoked as the `realUserName`, which returns its `geteuid()`.

Here I am returning `0` to indicate back to the test that everything worked successfully.

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 10 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: The reason for `OuterLoop` is so it doesn't run normally on dev machines.  We do this similar pattern elsewhere:  

https://github.com/dotnet/corefx/blob/4bf90198c5e7710ea67b7c1b467e859ec8bb78e5/src/System.IO.FileSystem/tests/File/Delete.cs#L127-L129
https://github.com/dotnet/corefx/blob/4bf90198c5e7710ea67b7c1b467e859ec8bb78e5/src/System.IO.FileSystem/tests/Directory/Delete.cs#L206-L208
https://github.com/dotnet/corefx/blob/bf412174f02f751445df31a6393417d4eece5a16/src/System.Net.Security/tests/FunctionalTests/NegotiateStreamTest.Linux.cs#L118  (all the tests are marked with `OuterLoop`)

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs (hunk 5 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: I copied the code from

https://github.com/dotnet/corefx/blob/17be4bc9b3fd9443026d2bf7d11a043b940bbcf4/src/System.Runtime.Extensions/src/System/Environment.Unix.cs#L429-L430
and
https://github.com/dotnet/corefx/blob/bfac45af5dddcbd70b6935fcf92b4b8755f3cd1c/src/Common/src/System/IO/PersistedFiles.Unix.cs#L159-L160

It seemed consistent to throw the same exception here.  But then looking at the Process.Start documentation, it lists Win32Exception and doesn't list IOException.  So I'll change to Win32Exception.

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs (hunk 8 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: Done.

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs (hunk 6 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: No, they are either both set, or both null.  I'll check just `userId` and add an Assert for `groupId`.

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs (hunk 8 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs]: Obviously it can't be `0`, since we double in the fallback path.  But I'll start it at `2` in DEBUG cases, which is not large enough to hold a normal user name.

File: src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs (hunk 12 above)
Comment (eerhardt) [reply src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs]: I've added this test.

**Comment by eerhardt** [comment]
Test NETFX x86 Release Build

**Comment by eerhardt** [comment]
Failures are flaky SQLClient tests. Ignoring.
